### DISCUSSION TITLE
Simple Webservice - remove CDATA from response

### DISCUSSION THREAD

**Discussion by FTK7** [nextgenhealthcare/connect#5173] [Q&A] [answered]
Hi, 
My question is the same in https://forums.mirthproject.io/forum/mirth-connect/support/17961-remove-cdata-wrap-in-web-service-listener-response, no answer there.

I have made a simple webservice in Mirth, in the source it just uses default webservice listener with acceptMessage:
![imatge](https://user-images.githubusercontent.com/66680696/166665354-06b4bc7e-cecd-4319-82ad-715695def3a2.png)

In the destination I have a step in "edit response" which fills two fields:
![imatge](https://user-images.githubusercontent.com/66680696/166665890-0c14cab2-8eda-43e3-9390-2dd992ab5095.png)
It is a javascript writer that only returns the data:
return(connectorMessage.getEncodedData());

Everything works ok but when I try it it answers the data inside a CDATA tag:
  ```
    <ns2:acceptMessageResponse>
         <return><![CDATA[<usuari><dni>DNI</dni><tel>777</tel></usuari>]]></return>
      </ns2:acceptMessageResponse>
```

I would like to recieve the xml code to read it easily. Is it possible to remove the CDATA tags in the response?

Thank you,
Albert

**Comment by jonbartels** [comment]
You have a similar question as https://github.com/nextgenhealthcare/connect/discussions/5168

Something like this (untested, bugs guaranteed!)
```
var myMessage = connectorMessage.getEncodedData();
//note the square bracket notation, return is a JS keyword this effectively escapes it
//e4x should just automatically read the CDATA as a string
var cdataBlock = myMessage.*::acceptMessageResponse.*::['return'].toString();
//the contents of CDATA are a string. parse it back to real XML
var myXml = new XML(cdataBlock);
//now we can read the XML dni node contents
var myDni = myXML.dni.toString();
```

PS - thank you for posting the forum link. If you find a solution in one venue, please cross post it to the other one. This helps other users!

**Comment by tonygermano** [comment, marked as answer]
CDATA is just an alternate way to escape a string (which is what acceptMessage returns.) If it did not add the CDATA, it would have to xml escape your response, because it is only expecting a String, and not child xml elements. The CDATA does make it easier to read than having `&lt;` all over the place.

**Reply by FTK7** [reply]
So, I guess, if I want to return child xml elements; should I overwrite the acceptMessage method? Or returning xml elements is not possible?
If I understand, acceptMessage only returns String, so Mirth automatically escapes the text returned.

**Reply by FTK7** [reply]
Rewriting acceptMessage method is the solution. As tonygermano said, it only returns a String, so you have to rewrite to return another object.
I did something like this (it returns an object pacient):
```
@WebService
public class WS_Savac extends AcceptMessage {
    public WS_Savac(WebServiceReceiver webServiceReceiver) {
		super(webServiceReceiver);
	}
    @WebMethod
    public Pacient dadesRegistre(@WebParam(name = "idUsuari") String idUsuari) {
        String response = webServiceReceiver.processData(idUsuari);
        Pacient pacient = new Pacient();

        if (response != null) {
        	 try {
	        		 JAXBContext jaxbContext = JAXBContext.newInstance(Pacient.class);
	        		 Unmarshaller unmarshaller = jaxbContext.createUnmarshaller();
	        		 StringReader reader = new StringReader(response);
	        		 pacient = (Pacient) unmarshaller.unmarshal(reader);
        		 } catch (JAXBException e) {
        		 e.printStackTrace();
        		 }
            return pacient;
        }
        
        return null;
    }
}
```

Pacient is a java class with 3 fields, dni, email and mobil:

```
@XmlRootElement(name = "pacient")
@XmlAccessorType(XmlAccessType.NONE) //Sense binding automatic
@XmlType(propOrder = { "dni", "email", "mobil" })
public class Pacient implements Serializable{
	 /**
	 * 
	 */
	private static final long serialVersionUID = 1663453105794674467L;
	private String dni;
	private String email;
	private String mobil;
	@XmlElement(name = "dni")
	public String getDni() {
		return dni;
	}
	@XmlElement(name = "email")
	public String getEmail() {
		return email;
	}
	@XmlElement(name = "mobil")
	public String getMobil() {
		return mobil;
	}
	 public void setDni(String newDni) {
		 dni = newDni;
	 }
	 public void setEmail(String newEmail) {
		 email = newEmail;
	 }
	 public void setMobil(String newMobil) {
		 mobil = newMobil;
	 }
}
```


For information, I found this example in French:
https://cboyer.github.io/developpement/mirth-webservice-soap/?_x_tr_sl=auto
